### PR TITLE
fix: stop embedding tracker PATs in git remote URLs (#173)

### LINKED ISSUE #173: [Bug] GitHub PAT embedded in clone URL is leaked via flow execution logs (git remote -v)
## Summary

When a flow clones a private GitHub repo using a tracker PAT, Preloop embeds the token in the remote URL (`https://github_pat_…@github.com/org/repo.git`). Agents commonly run `git remote -v` (or Preloop itself logs remotes). That prints the **full PAT into flow execution logs**, which are visible in the Preloop UI/API.

This is a credential disclosure bug. We had to revoke and rotate fine-grained PATs after both a Pull Request Reviewer run and a Security Vulnerability Scanner run leaked tokens this way.

## Steps To Reproduce

1. Configure a GitHub tracker with a fine-grained PAT (Contents: Read, and optionally Pull requests: Read/write).
2. Enable `git_clone_config` on a flow so Preloop clones from that tracker (e.g. Pull Request Reviewer or Security Vulnerability Scanner).
3. Trigger the flow against a private repo (real webhook or manual trigger with repository payload).
4. After clone succeeds, either:
   - Let the Codex/agent run inspect remotes (`git remote -v`), or
   - Inspect the execution logs / `model_output_summary` via the API/UI.
5. Observe log lines of the form:

```text
origin  https://github_pat_<FULL_TOKEN>@github.com/<org>/<repo>.git (fetch)
origin  https://github_pat_<FULL_TOKEN>@github.com/<org>/<repo>.git (push)
```

## Expected Behavior

- Tracker credentials must never appear in execution logs, agent transcripts, or `model_output_summary`.
- Prefer a non-logging credential helper / `GIT_ASKPASS` / insteadOf rewrite so the working tree remote is a clean `https://github.com/org/repo.git` without an embedded secret.
- If a secret is detected in outbound log lines, redact (`github_pat_[REDACTED]`, `ghp_[REDACTED]`, etc.) before persistence and UI display.
- Consider scrubbing existing retained logs that may already contain leaked tokens.

## Environment

- Preloop: Cloud (`https://preloop.ai`)
- CLI: `0.13.1` (commit `3eb7341`)
- Flows: Pull Request Reviewer, Security Vulnerability Scanner (Codex, ephemeral)
- Tracker: GitHub org private repos
- Auth form observed in remotes: `https://github_pat_…@github.com/...` (token as URL username)

## Additional Context

- Impact: any operator or API client who can read flow execution logs can exfiltrate the tracker PAT; logs may also land in support exports, backups, or SIEM.
- Workaround: rotate PATs after each leak; avoid instructing agents to run `git remote -v` (insufficient — the platform still must not log secrets).
- Related: openai-compatible OpenRouter model listing / LiteLLM auto-beta gateway issues filed separately; this is independent credential handling.

### PR DESCRIPTION
**Pull request by d-mo**
Target release: **v0.14.0**. Fixes #173.

## The leak

Flows with `git_clone_config` had the tracker PAT spliced into the clone URL:

```
https://github_pat_xxxx@github.com/org/repo.git
```

That URL becomes the cloned repository's `origin` remote, so any `git remote -v`
the agent ran printed the token straight into the flow execution log, which is
served over the API and rendered in the console. The reporter had rotated PATs
twice before the underlying cause was identified.

## Primary fix: tokens never touch URLs

New `preloop.utils.git_credentials` hands the token to git through a credential
store helper instead:

1. The token travels to the container as `PRELOOP_GIT_CREDENTIALS`.
2. The init script writes it to a `0600` file under `umask 077`, points
   `credential.helper` at that file, then **unsets the variable**, so the agent
   process cannot read it back out of its own environment.
3. The URL handed to `git clone` is always credential-free, so the remote it
   writes into `.git/config` is safe to print.

The secret is deliberately never rendered into the shell script itself, since
some images echo their init commands and the script is visible in
`kubectl describe`.

`GIT_TERMINAL_PROMPT=0` is set so a credential miss fails fast instead of
hanging the container until timeout.

## Five sites embedded credentials, not one

The issue pointed at the clone path. Auditing turned up five:

| Site | Fix |
| --- | --- |
| `container.py::_inject_git_credentials_into_url` | removed |
| `container.py::_get_repo_url_from_project` | returns a credential-free URL |
| `container.py` post-exec PR/MR curls | raw token was interpolated into the generated script; now reads `${PRELOOP_GIT_TOKEN_N}` |
| `openhands.py::_prepare_git_clone_command` | has its own clone path, same fix |
| `flow_orchestrator.py::_inject_credentials_into_url` | removed; clone now runs with an ephemeral credential file unlinked in a `finally` |

`repo_urls.inject_oauth_token` is **deleted** rather than deprecated, so it
cannot be reintroduced by autocomplete. A comment in its place explains why.

## Defense in depth: log scrubbing

New `preloop.utils.secret_scrubbing` redacts:

- URL userinfo (`https://user:token@host` and `https://token@host`), with an
  allowlist so harmless usernames like `git` and `x-access-token` survive and
  logs stay readable
- `Authorization: Bearer/Token/Basic`, `PRIVATE-TOKEN:`, `X-Gitlab-Token:`
- Known token formats: `github_pat_`, `ghp_`/`gho_`/`ghu_`/`ghs_`/`ghr_`,
  `glpat-`, GitLab project/deploy tokens, `sk-ant-`, `sk-`, `preloop_`

Each replacement keeps its prefix, so a redacted log still tells an operator
which kind of credential leaked and therefore what to rotate.

Applied at four layers:

1. `container.get_logs` (read)
2. `container.stream_logs` (live console feed)
3. `FlowExecutionLogger.log_agent_output`, which feeds `model_output_summary`,
   named explicitly in the issue
4. **both** `append_log` implementations, as the last gate before persistence,
   covering producers that skip scrubbing at the source

## Incidental fixes

`repo_url` and `branch` were interpolated unquoted into the openhands and
orchestrator clone commands. Both are webhook-derived. Now `shlex.quote`d, with
tests.

## Testing

100 new tests. Highlights:

- A real-git integration test that runs the generated setup script through a
  shell, clones, and asserts `git remote -v` and `.git/config` contain no token,
  that the credential file is `0600`, and that a later `env` dump cannot see the
  secret.
- Redaction tests built from the leak samples in the issue, plus false-positive
  tests so email addresses, SSH remotes, commit SHAs and clean URLs pass through
  untouched.
- `append_log` tests parameterized over both CRUD implementations.

Verified by mutation: making `strip_url_credentials` a no-op fails 7 tests
across all four affected layers, and disabling `scrub_secrets` fails 36.

Full backend suite: **229 pre-existing failures before and after, identical
sets** (they stem from a local `.env` whose `SECRET_KEY` cannot decrypt test
fixtures, and reproduce on a clean worktree at the base commit). No new mypy
errors; both new modules are mypy-clean.

## Not included

This stops new leaks. Logs already retained from before the fix may still
contain tokens; no backfill is performed here. Worth a follow-up issue, since
the whole point of the report is that a rotated token still sits in history.

Related but untouched, per the brief: the 7 errored agent pods in
preloop-prod are on this same code path. **Prod was not touched.**

> [!NOTE]
> **Low Risk**
> Defense-in-depth security fix: tokens leave URLs and get 4-layer log scrubbing. Well-tested with real-git integration tests and mutation verification. No new mypy errors.
>
> **Overview**
> Fixes credential leak (#173) by moving PATs from git clone URLs into a credential helper, scrubbing agent logs at every persistence layer, and hardening shell commands. 100+ tests cover the fix, edge cases, and false-positive guardrails.
>
> <sup>Written by [Preloop PR Reviewer](https://preloop.ai) for commit 38434c5. Updates automatically on new commits.</sup>

### CODE REVIEW

File: backend/preloop/utils/git_credentials.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        env = os.environ.copy()
+        # -c flags apply to the git invocation without mutating any config file.
+        env["GIT_CONFIG_COUNT"] = "1"
+        env["GIT_CONFIG_KEY_0"] = "credential.helper"
+        env["GIT_CONFIG_VALUE_0"] = f"store --file={path}"
+        env["GIT_TERMINAL_PROMPT"] = "0"
+        yield env
+    finally:
+        try:
+            os.unlink(path)
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/preloop/utils/git_credentials.py:248]: Narrowed to `except FileNotFoundError` in d418fc46, so permission and I/O errors surface instead of being swallowed.

File: backend/preloop/utils/secret_scrubbing.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            lambda m: f"{m.group('prefix')}{REDACTED}", scrubbed
+        )
+
+    for pattern, replacement in _TOKEN_PATTERNS:
+        scrubbed = pattern.sub(replacement, scrubbed)
+
+    return scrubbed
+
+
+def scrub_secret_lines(lines: List[str]) -> List[str]:
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/preloop/utils/secret_scrubbing.py:128]: Kept the `or ""` and documented the intent in d418fc46: it is a deliberate None-safety guard so callers always get a list of strings.

File: backend/preloop/services/flow_execution_logger.py [hunk 3]
```diff
@@ -147,10 +148,14 @@ def log_agent_output(self, line: str):
...
 
+        The line is scrubbed of known credential formats before it is stored,
+        because these lines become ``model_output_summary``, which is served
+        over the API and rendered in the console (issue #173).
+
         Args:
             line: A single line of agent stdout/stderr
         """
-        self.agent_output_lines.append(line)
+        self.agent_output_lines.append(scrub_secrets(line) or "")
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/preloop/services/flow_execution_logger.py:158]: Added a lock-in test in d418fc46 (`test_token_only_line_keeps_redaction_marker`): a token-only line must scrub to `github_pat_[REDACTED]`, never the empty string. Verified by mutation, emptying the replacement makes the test fail.

File: backend/preloop/services/flow_orchestrator.py [hunk 4]
```diff
@@ -964,40 +972,47 @@
...
 
-            stdout, stderr = await process.communicate()
+                stdout, stderr = await process.communicate()
 
             if process.returncode != 0:
                 logger.error(
-                    f"Git clone failed with code {process.returncode}: {stderr.decode()}"
+                    "Git clone failed with code %s: %s",
+                    process.returncode,
+                    scrub_secrets(stderr.decode()),
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/preloop/services/flow_orchestrator.py:1011]: Dismissed as a false positive (code scanning alert). The flagged expression is the sanitizer itself: subprocess output passes through `scrub_secrets()` before it reaches the logger. CodeQL does not model `scrub_secrets` as a sanitizer. PR #178 adds 4-layer scrubbing with mutation tests, disabling `scrub_secrets` fails 36 tests.

File: backend/preloop/services/flow_orchestrator.py [hunk 5]
```diff
@@ -964,40 +972,47 @@
...
                 )
                 return None
 
-            logger.info(f"Git clone successful: {stdout.decode()}")
+            logger.info(f"Git clone successful: {scrub_secrets(stdout.decode())}")
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/preloop/services/flow_orchestrator.py:1015]: Dismissed as a false positive (code scanning alert). The flagged expression is the sanitizer itself: subprocess output passes through `scrub_secrets()` before it reaches the logger. CodeQL does not model `scrub_secrets` as a sanitizer. PR #178 adds 4-layer scrubbing with mutation tests, disabling `scrub_secrets` fails 36 tests.

File: backend/tests/utils/test_repo_urls.py [hunk 6]
```diff
@@ -21,31 +22,24 @@ def test_rejects_substring_false_positives(self) -> None:
...
+    It embedded the tracker token in the clone URL, which made the secret part
+    of the repository's ``origin`` remote and leaked it through ``git remote
+    -v`` into flow execution logs. Credentials are now delivered by
+    ``preloop.utils.git_credentials`` via a git credential helper.
+
+    This test pins the removal so the helper cannot quietly come back.
+    """
+
+    def test_helper_is_gone(self) -> None:
+        import preloop.utils.repo_urls as repo_urls
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/tests/utils/test_repo_urls.py]: Fixed in d418fc46: the file now uses a single module-level import of `repo_urls` instead of mixing import styles.

File: backend/tests/services/test_flow_orchestrator_git_credentials.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    async def _run_clone(self, orchestrator, tracker_credentials):
+        captured = {}
+
+        async def fake_subprocess_shell(cmd, **kwargs):
+            captured["cmd"] = cmd
+            captured["env"] = kwargs.get("env")
+            if captured["env"] and "GIT_CONFIG_VALUE_0" in captured["env"]:
+                path = captured["env"]["GIT_CONFIG_VALUE_0"].split("=", 1)[1]
+                captured["credential_file"] = path
+                captured["credential_file_content"] = open(path).read()
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/tests/services/test_flow_orchestrator_git_credentials.py]: Fixed in d418fc46: the credential file is now read via a `with open(...)` context manager.

File: backend/tests/utils/test_git_credentials.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+        credential = GitCredential(
+            "https://github.com/acme/private.git", "x-access-token", LEAKED_PAT
+        )
+        with temporary_credential_file(credential) as env:
+            assert env is not None
+            match = re.search(r"store --file=(\S+)", env["GIT_CONFIG_VALUE_0"])
+            assert match
+            path = match.group(1)
+
+            assert LEAKED_PAT in open(path).read()
```
[the remark this answers is not in the document]
Comment (d-mo) [reply backend/tests/utils/test_git_credentials.py]: Fixed in d418fc46: the read now happens in a `with` block and the assert is side-effect free (`assert LEAKED_PAT in content`).

[the remark this answers is not in the document]
Comment (d-mo) [reply backend/tests/utils/test_git_credentials.py]: Fixed in d418fc46: same site, the file is read via a context manager and the side-effecting read moved out of the assert.
